lib/syntax.ml

```diff
@@ -205,43 +212,51 @@ let rec resolve_columns env expr =
       let schema = Schema.Source.from_schema schema in
       (* represet nested selects as functions with sql parameters as function arguments, some hack *)
       match schema, usage with
-      | [ {domain;_} ], `AsValue -> ResFun (Type.Ret domain.t, as_params p) (* use nullable *)
+      | [ {domain;_} ], `AsValue -> ResSelectExpr (Type.Ret domain.t, as_params p) (* use nullable *)
       | s, `AsValue -> raise (Schema.Error (s, "only one column allowed for SELECT operator in this expression"))
-      | _, `Exists -> ResFun (Type.Ret Any, as_params p)
+      | _, `Exists -> ResSelectExpr (Type.Ret Any, as_params p)
   in
   each expr
 
 (** assign types to parameters where possible *)
 and assign_types env expr =
   let { set_tyvar_strict; _ } = env in
   let option_split = function None -> None, None | Some (x,y) -> Some x, Some y in
-  let rec typeof_ (e:res_expr) = (* FIXME simplify *)
+  let rec typeof_ ~under_agg (e:res_expr) = (* FIXME simplify *)
     match e with
     | ResValue t -> e, `Ok t
     | ResParam p -> e, `Ok p.typ
     | ResInparam p -> e, `Ok p.typ
     | ResInTupleList (param_id, kind) -> 
       (match kind with 
       | Res res_exprs -> ResInTupleList (param_id, ResTyped (List.map (fun expr ->
-         let typ = expr |> typeof |> snd |> get_or_failwith in 
+         let typ = expr |> typeof ~under_agg |> snd |> get_or_failwith in 
          if Type.is_any typ then 
             fail "If you need to have a field as parameter in the left part you should specify a type"
          else typ
         ) res_exprs)), `Ok (Type.strict Bool) 
       | ResTyped _ -> assert false
       )
-    | ResInChoice (n, k, e) -> let e, t = typeof e in ResInChoice (n, k, e), t
+    | ResInChoice (n, k, e) -> let e, t = typeof ~under_agg e in ResInChoice (n, k, e), t
     | ResChoices (n,l) ->
-      let (e,t) = List.split @@ List.map (fun (_,e) -> option_split @@ Option.map typeof e) l in
+      let (e,t) = List.split @@ List.map (fun (_,e) -> option_split @@ Option.map (typeof ~under_agg) e) l in
       let t =
         match Type.common_subtype @@ List.map get_or_failwith @@ List.filter_map identity t with
         | None -> `Error "no common subtype for all choice branches"
         | Some t -> `Ok t
       in
       ResChoices (n, List.map2 (fun (n,_) e -> n,e) l e), t
+    | ResSelectExpr(func, params) -> 
+      let (res, typ) = typeof ~under_agg (ResFun (func, params)) in 
+      let typ = get_or_failwith typ in
+      (* Any nested query under aggregation is a possible no-rows result that returns as NULL *)
```

Comment:
this feels strange, why nullability decision is made here? nested query can return no rows, yes, but then the caller which is applying aggregation shall mark it as nullable, not here

Reply:
If I understood correctly what you mean then, lets look at this example:

```sql
SELECT MAX(COALESCE(((SELECT value FROM test18 WHERE value = 100)), 1)) AS result
    FROM test18
    GROUP BY value
```

In this case

> the caller which is applying aggregation

is `MAX`

And we do

> shall mark

We say that `COALESCE` is nullable, but in this case shouldn't return null.

And I need a recursive pass here, maybe I'm wrong. I'll try to move the logic up anyway. But it seems that this current solution, although it changes the `type_of` signature, is a simpler solution

Reply:
Fixed, I made a new pass for an aggregation